subsys/lorawan/lorawan.c

```diff
@@ -0,0 +1,601 @@
+#elif CONFIG_LORAMAC_REGION_CN779
+	#define LORAWAN_REGION LORAMAC_REGION_CN779
+#elif CONFIG_LORAMAC_REGION_EU433
+	#define LORAWAN_REGION LORAMAC_REGION_EU433
+#elif CONFIG_LORAMAC_REGION_EU868
+	#define LORAWAN_REGION LORAMAC_REGION_EU868
+#elif CONFIG_LORAMAC_REGION_KR920
+	#define LORAWAN_REGION LORAMAC_REGION_KR920
+#elif CONFIG_LORAMAC_REGION_IN865
+	#define LORAWAN_REGION LORAMAC_REGION_IN865
+#elif CONFIG_LORAMAC_REGION_US915
+	#define LORAWAN_REGION LORAMAC_REGION_US915
+#elif CONFIG_LORAMAC_REGION_RU864
+	#define LORAWAN_REGION LORAMAC_REGION_RU864
+#elif
+	#error "Atleast one LoRaWAN region should be selected"
+#endif
+
+/* Use version 1.0.3.0 for ABP */
+#define LORAWAN_ABP_VERSION 0x01000300
+
+#define LOG_LEVEL CONFIG_LORAWAN_LOG_LEVEL
+#include <logging/log.h>
+LOG_MODULE_REGISTER(lorawan);
+
+K_SEM_DEFINE(mlme_confirm_sem, 0, 1);
+K_SEM_DEFINE(mcps_confirm_sem, 0, 1);
+
+K_MUTEX_DEFINE(lorawan_join_mutex);
+K_MUTEX_DEFINE(lorawan_send_mutex);
+
+static enum lorawan_datarate lorawan_datarate = LORAWAN_DR_0;
+static uint8_t lorawan_conf_msg_tries = 1;
+static bool lorawan_adr_enable;
+
+const char *status2str(int status)
+{
+	switch (status) {
+	case LORAMAC_STATUS_OK:
+		return "OK";
+	case LORAMAC_STATUS_BUSY:
+		return "Busy";
+	case LORAMAC_STATUS_SERVICE_UNKNOWN:
+		return "Service unknown";
+	case LORAMAC_STATUS_PARAMETER_INVALID:
+		return "Parameter invalid";
+	case LORAMAC_STATUS_FREQUENCY_INVALID:
+		return "Frequency invalid";
+	case LORAMAC_STATUS_DATARATE_INVALID:
+		return "Datarate invalid";
+	case LORAMAC_STATUS_FREQ_AND_DR_INVALID:
+		return "Frequency or datarate invalid";
+	case LORAMAC_STATUS_NO_NETWORK_JOINED:
+		return "No network joined";
+	case LORAMAC_STATUS_LENGTH_ERROR:
+		return "Length error";
+	case LORAMAC_STATUS_REGION_NOT_SUPPORTED:
+		return "Region not supported";
+	case LORAMAC_STATUS_SKIPPED_APP_DATA:
+		return "Skipped APP data";
+	case LORAMAC_STATUS_DUTYCYCLE_RESTRICTED:
+		return "Duty-cycle restricted";
+	case LORAMAC_STATUS_NO_CHANNEL_FOUND:
+		return "No channel found";
+	case LORAMAC_STATUS_NO_FREE_CHANNEL_FOUND:
+		return "No free channel found";
+	case LORAMAC_STATUS_BUSY_BEACON_RESERVED_TIME:
+		return "Busy beacon reserved time";
+	case LORAMAC_STATUS_BUSY_PING_SLOT_WINDOW_TIME:
+		return "Busy ping-slot window time";
+	case LORAMAC_STATUS_BUSY_UPLINK_COLLISION:
+		return "Busy uplink collision";
+	case LORAMAC_STATUS_CRYPTO_ERROR:
+		return "Crypto error";
+	case LORAMAC_STATUS_FCNT_HANDLER_ERROR:
+		return "FCnt handler error";
+	case LORAMAC_STATUS_MAC_COMMAD_ERROR:
+		return "MAC command error";
+	case LORAMAC_STATUS_CLASS_B_ERROR:
+		return "ClassB error";
+	case LORAMAC_STATUS_CONFIRM_QUEUE_ERROR:
+		return "Confirm queue error";
+	case LORAMAC_STATUS_MC_GROUP_UNDEFINED:
+		return "Multicast group undefined";
+	case LORAMAC_STATUS_ERROR:
+		return "Unknown error";
+	default:
+		return NULL;
+	}
+}
+
+const char *eventinfo2str(int status)
```

Comment:
Is this function going to be used outside of lorawan.c?

Reply:
They are all using MAC-internal status codes, so they can't be used from outside of the driver. However, we will probably need to use them from other source files when adding more functionality. I ended up with [this change](https://github.com/andysan/zephyr/commit/846adf11af2cc71bbcbef18c86cb7e14036514a5) when implementing [state storage](https://github.com/andysan/zephyr/commits/lw_persistent_state).